ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixAggregatedViewClusterInitializer.java

```diff
@@ -0,0 +1,198 @@
+   */
+  HelixAggregatedViewClusterInfo join() throws Exception {
+    try {
+      return initializationFuture.get();
+    } catch (ExecutionException e) {
+      throw Utils.extractFutureExceptionCause(e);
+    }
+  }
+
+  /**
+   * Callback to complete the future when an error occurs
+   * @param e the {@link Exception}.
+   */
+  private void onInitializationFailure(Exception e) {
+    initializationFuture.completeExceptionally(e);
+  }
+
+  /**
+   * Perform initialization steps related to registering with Helix Aggregated view cluster.
+   * @return the {@link HelixAggregatedViewClusterInfo} for the cluster.
+   * @throws Exception if something went wrong during startup
+   */
+  public HelixAggregatedViewClusterInfo initializeHelixAggregatedViewCluster() throws Exception {
+    // We can assume that Helix Aggregated view cluster will be present in same zookeeper hosting regular Ambry cluster
+    // information. Get the HelixManager to talk to the local zk service.
+    DcZkInfo localDcZkInfo = dataCenterToZkAddress.get(clusterMapConfig.clusterMapDatacenterName);
+
+    // For now, the first ZK endpoint (if there are more than one endpoints) will be adopted by default for initialization.
+    // Note that, Ambry currently doesn't support multiple spectators, because there should be only one source of truth.
+    String localZkConnectStr = localDcZkInfo.getZkConnectStrs().get(0);
+    HelixManager aggregatedViewManager =
+        helixFactory.getZkHelixManagerAndConnect(clusterMapConfig.clusterMapAggregatedViewClusterName, selfInstanceName,
+            InstanceType.SPECTATOR, localZkConnectStr);
+    logger.info("Helix cluster name {}", aggregatedViewManager.getClusterName());
+
+    HelixClusterChangeHandler clusterChangeHandler =
+        helixClusterManager.new HelixClusterChangeHandler(clusterMapConfig.clusterMapClusterName,
+            this::onInitializationFailure, true);
+
+    // Helix aggregated view cluster currently aggregates the following:
+    // 1. External view (which tells the current state of the cluster like replicas and their states).
+    // 2. Instance configs (These mostly contain fields set and used by Helix like host name and port).
+    // 3. Live instances (List of live hosts in cluster).
+
+    // Ambry currently needs and listens to following changes in Helix/ZKs in all datacenters:
+    // 1. External view (indirectly via using RoutingTableProvider)
+    // 2. Live instances (List of live hosts in cluster)
+    // 3. HelixPropertyStore. This is where Ambry data node configuration like disk information, disk to replica mapping,
+    //    replica sealed states, etc are stored.
+    // 4. Helix ideal states. This is needed to keep track of Helix Resource (Helix terminology) to Ambry partitions
+    //    (Multiple Ambry partitions are grouped under one helix resource).
+
+    // As we can see above, since #3 and #4 are not aggregated by Helix currently, we get these information by talking
```

Comment:
Can we ask the Helix team to bring #3 and #4 to the aggregate view too?  If we still need to talk to all ZK in all DC, how much improvement are we getting?

Reply:
Sure, will check with them. Based on earlier discussion with them, it looks like one of the reasons property store (#3) is not aggregated right now is that it has custom application data (and not helix data). So the folder structure/fields is probably not guaranteed to be same across all data centers. And helix team wanted to know on why we needed to aggregate ideal states (#4).

Right now, 70% of time is spent in fetching item routing table #1. So, we are still getting 70% improvement.

Reply:
I am quite surprised that 70% of the time is spent on fetching routing table. Because routing table is the external view and external view has same amount of data to fetch comparing to IdealState. Why is that fetching IdealState is fast, but external view is slow?

Reply:
My understanding is that for getting most up-to-date external view, helix would still have to go from host to host, check list of live hosts, etc and aggregate the current states of partitions in each host to generate external view. Due to that generation of external view takes time. 

Also, based on my discussion with helix team, it seems like there are two modes of instantiating a Routing Table. One is EXTERNAL_VIEW based and other is CURRENT_STATES based. For the former one, calculation of external view is done by helix controller and we wait until the most up-to-date view is generated. But, it has less read traffic (almost same data as ideal states) but has dependency on helix controller and is said to take longer time. For the latter one, calculation of external view (by going from host to host and getting partition states, etc) is done at the helix agent. Due to that, we avoid dependency on helix controller (and is supposed to be faster) but has more read traffic. We seem to use CURRENT_STATES based Routing Table. Even this takes time since we need to read all the data from remote Zookeepers currently.